### try:异常处理

**1.单支**

```python
try:
	a = int(input('请输入数字'))
except ValueError:
	print('单支报错')
```

**2.多支**

```python
try:
	a = int(input('请输入数字'))
except ValueError:
	print('请输入数字')
except NameError:
	print('请给变量名赋予值')
```

**3.万能,任何错误都可以捕捉(特别除外)**

```python
 s1 = 'dasdads'
try:
	int(s1)
except Exception as e:
	print(e)
```

**4.万能+多支:万能需要放最后**

```python
s1 = 'asdadad'
try:
	int(s1)
except IndexError as e:
	print(e)
except KeyError as e:
	print(e)
except ValueError as e:
	print(e)
except Exception as e:
	print(e)
```

**5.异常的其他机构'**

```python
s1 = 'asdadad'
try:
	int(s1)
except IndexError as e:
	print(e)
except KeyError as e:
	print(e)
except ValueError as e:
	print(e)
except Exception as e:
	print(e)
else:
	print('try里面的代码没有问题就执行我')
finally:
	print('不管有什么错误,我都一定会执行的')
```

**6.主动触发**

```python
try:
	raise TypeError('类型错误')
except Exception as e:
	print(e)
```

**7.自定义异常**

```python
class EvaException(BaseException):
	def __init__(self,msg):
		self.msg = msg
	def __str__(self):
		return self.msg
try:
	raise EvaException('类型错误')
except EvaException as e:
	print(e)
```

**8.断言**:只要**assert**后面的条件不满足,就会报错,下面的一切代码都不会执行 

```python
print('asdadasdasd')
assert 1==2
print('asdadads')
```

**9.try..except的方式比较if的方式的好处** 

- 1.把错误处理和真正的工作分开来
- 2.代码更容易组织了,更清晰,复杂的工作任务更容易实现
- 3.更安全了,不至于由于一些小的疏忽而使程序意外崩溃了

### 异常报错总结

- 1.**AttributeError** 试图访问一个对象没有的树形，比如foo.x，但是foo没有属性x
- 2.**IOError** 输入/输出异常；基本上是无法打开文件
- 3.**ImportError** 无法引入模块或包；基本上是路径问题或名称错误
- 4.**IndentationError** 语法错误（的子类） ；代码没有正确对齐
- 5.**IndexError** 下标索引超出序列边界，比如当x只有三个元素，却试图访问x[5]
- 6.**KeyError** 试图访问字典里不存在的键
- 7.**KeyboardInterrupt** Ctrl+C被按下
- 8.**NameError** 使用一个还未被赋予对象的变量
- 9.**SyntaxError** Python代码非法，代码不能编译(个人认为这是语法错误，写错了）
- 10.**TypeError** 传入对象类型与要求的不符合
- 11.**UnboundLocalError** 试图访问一个还未被设置的局部变量，基本上是由于另有一个同名的全局变量， 导致你以为正在访问它
- 12.**ValueError** 传入一个调用者不期望的值，即使值的类型是正确的

### 日常报错总结

